<a href="https://colab.research.google.com/github/lustraka/Data_Analysis_Workouts/blob/main/Analyse_Twitter_Data/practice-assessing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Oral Insulin Phase II Clinical Trial Data

This project looks at the phase two clinical data of 350 patients for a new innovative oral insulin called Auralin - a proprietary capsule that can solve the stomach lining problem.

Phase two trials are intended to:
- Test the efficacy and the dose response of a drug
- Idenitify adverse reactions

In this trial, half of the patients are being treated with Auralin, and the other 175 being treated with a popular injectable insulin called Novodra. By comparing key metrics between these two drugs, we can determine if Auralin is effective.

*DISCLAIMER: This data isn't real. It was fabricated for the sake of this project.*

## Data Quality Dimensions (Aspects, Measures, Categories)
Data quality dimensions help guide your thought process while assessing and also cleaning. The four main data quality dimensions are:

1. **Completeness**: do we have all of the records that we should? Do we have missing records or not? Are there specific rows, columns, or cells missing?
2. **Validity**: we have the records, but they're not valid, i.e., they don't conform to a defined schema. A schema is a defined set of rules for data. These rules can be real-world constraints (e.g. negative height is impossible) and table-specific constraints (e.g. unique key constraints in tables).
3. **Accuracy**: inaccurate data is wrong data that is valid. It adheres to the defined schema, but it is still incorrect. Example: a patient's weight that is 5 lbs too heavy because the scale was faulty.
4. **Consistency**: inconsistent data is both valid and accurate, but there are multiple correct ways of referring to the same thing. Consistency, i.e., a standard format, in columns that represent the same data across tables and/or within tables is desired.
These are listed in decreasing order of severity, meaning that the dimension listed first, completeness, is the most important.

Regarding the other data quality research mentioned in the video, the additional dimensions are specific cases of these four dimensions listed above. Example: *currency*, defined as follows: *the degree to which data is current with the world that it models. Currency can measure how up-to-date data is.* Currency is a specific case of accuracy in data in the sense that out-of-date data is (usually) valid but wrong. In other words, our definition of accuracy can include currency.

References:
 - [Data Strategy: How to Improve Data Quality](https://www.informit.com/articles/article.aspx?p=399325&seqNum=3)
 - [YouTube, Anish Raivadera: What Are The Seven Dimensions of Data Quality?](https://www.youtube.com/watch?v=dPsx8_Fcr-U)

## Data Tidiness
References:
- [Hadley Wickham: Tidy data](https://cran.r-project.org/web/packages/tidyr/vignettes/tidy-data.html)

## Gather

This Auralin Phase II clinical trial dataset comes in three tables: `patients`, `treatments`, and `adverse_reactions`.

In [10]:
import pandas as pd
import numpy as np
path = 'https://raw.githubusercontent.com/lustraka/Data_Analysis_Workouts/main/Analyse_Twitter_Data/data/'

In [2]:
patients = pd.read_csv(path + 'patients.csv')
treatments = pd.read_csv(path + 'treatments.csv')
adverse_reactions = pd.read_csv(path + 'adverse_reactions.csv')

## Assess
### The `patients` Table

In [3]:
patients
# For non-directed assessment also: df.sample(m)

patient_id assigned_sex given_name  ... weight height   bmi
0             1       female        Zoe  ...  121.7     66  19.6
1             2       female     Pamela  ...  118.8     66  19.2
2             3         male        Jae  ...  177.8     71  24.8
3             4         male       Liêm  ...  220.9     70  31.7
4             5         male        Tim  ...  192.3     27  26.1
..          ...          ...        ...  ...    ...    ...   ...
498         499         male    Mustafa  ...  181.1     72  24.6
499         500         male      Ruman  ...  239.6     70  34.4
500         501       female      Jinke  ...  171.2     67  26.8
501         502       female    Chidalu  ...  176.9     67  27.7
502         503         male        Pat  ...  138.2     71  19.3

[503 rows x 14 columns]

`patients` columns:
- **patient_id**: the unique identifier for each patient in the [Master Patient Index](https://en.wikipedia.org/wiki/Enterprise_master_patient_index) (i.e. patient database) of the pharmaceutical company that is producing Auralin
- **assigned_sex**: the assigned sex of each patient at birth (male or female)
- **given_name**: the given name (i.e. first name) of each patient
- **surname**: the surname (i.e. last name) of each patient
- **address**: the main address for each patient
- **city**: the corresponding city for the main address of each patient
- **state**: the corresponding state for the main address of each patient
- **zip_code**: the corresponding zip code for the main address of each patient
- **country**: the corresponding country for the main address of each patient (all United states for this clinical trial)
- **contact**: phone number and email information for each patient
- **birthdate**: the date of birth of each patient (month/day/year). The [inclusion criteria](https://en.wikipedia.org/wiki/Inclusion_and_exclusion_criteria) for this clinical trial is  age >= 18 *(there is no maximum age because diabetes is a [growing problem](http://www.diabetes.co.uk/diabetes-and-the-elderly.html) among the elderly population)*
- **weight**: the weight of each patient in pounds (lbs)
- **height**: the height of each patient in inches (in)
- **bmi**: the Body Mass Index (BMI) of each patient. BMI is a simple calculation using a person's height and weight. The formula is BMI = kg/m<sup>2</sup> where kg is a person's weight in kilograms and m<sup>2</sup> is their height in metres squared. A BMI of 25.0 or more is overweight, while the healthy range is 18.5 to 24.9. *The [inclusion criteria](https://en.wikipedia.org/wiki/Inclusion_and_exclusion_criteria) for this clinical trial is 16 >= BMI >= 38.*

In [11]:
# Date of birth as month/day/year instead of day/mont/year
np.array([int(d.split('/')[0]) for d in patients.birthdate.values]).max()

12

In [15]:
# The given_name for the patient with the patient_id 9 is Dsvid instead of David
patients.query('patient_id == 9')

patient_id assigned_sex given_name  ... weight height   bmi
8           9         male      Dsvid  ...  163.9     66  26.5

[1 rows x 14 columns]

In [16]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [20]:
# Select the records for patients that are form the city New York
patients.query('city == "New York"').shape

(18, 14)

In [24]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [28]:
patients.surname.value_counts()

Doe            6
Taylor         3
Jakobsen       3
Ogochukwu      2
Lund           2
              ..
Tomaszewski    1
Authier        1
Day            1
Leroux         1
Miles          1
Name: surname, Length: 466, dtype: int64

In [34]:
patients.query('surname == "Jakobsen"')

patient_id assigned_sex given_name  ... weight height   bmi
24           25         male      Jakob  ...  155.8     67  24.4
29           30         male       Jake  ...  155.8     67  24.4
432         433       female      Karen  ...  185.2     67  29.0

[3 rows x 14 columns]

In [35]:
patients.query('surname == "Knudsen"')

patient_id assigned_sex given_name  ... weight height   bmi
230         231       female  Elisabeth  ...  165.9     63  29.4

[1 rows x 14 columns]

In [29]:
patients.address.value_counts()

123 Main Street            6
648 Old Dear Lane          2
2476 Fulton Street         2
2778 North Avenue          2
3574 Park Boulevard        1
                          ..
3945 Simons Hollow Road    1
2690 Pin Oak Drive         1
4018 Poplar Avenue         1
932 Memory Lane            1
4220 Simpson Square        1
Name: address, Length: 483, dtype: int64

In [27]:
patients.loc[patients.address.duplicated()]

patient_id assigned_sex given_name  ... weight height   bmi
29           30         male       Jake  ...  155.8     67  24.4
219         220         male         Mỹ  ...  237.8     69  35.1
229         230         male       John  ...  180.0     72  24.4
230         231       female  Elisabeth  ...  165.9     63  29.4
234         235       female    Martina  ...  199.5     65  33.2
237         238         male       John  ...  180.0     72  24.4
242         243         male       John  ...  205.3     74  26.4
244         245         male       John  ...  180.0     72  24.4
249         250         male   Benjamin  ...  146.5     69  21.6
251         252         male       John  ...  180.0     72  24.4
257         258         male        Jin  ...  231.7     69  34.2
264         265       female   Wafiyyah  ...  158.6     63  28.1
269         270       female     Flavia  ...  175.2     61  33.1
277         278         male       John  ...  180.0     72  24.4
278         279       female   Generosa  ...  124.3     69  18.4
282         283       female      Sandy  ...  206.1     64  35.4
286         287         male      Lewis  ...  155.3     68  23.6
296         297       female        Chỉ  ...  181.1     63  32.1
502         503         male        Pat  ...  138.2     71  19.3

[19 rows x 14 columns]

In [30]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
       ...  
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [37]:
# The minimum value of weight is 48.8 pounds which looks very low
# Further examination shows that 48.8 is accually kilograms instead of pounds
weight_lbs = patients[patients.surname == 'Zaitseva'].weight * 2.20462
height_in = patients[patients.surname == 'Zaitseva'].height
bmi_check = 703 * weight_lbs / ( height_in * height_in )
print(bmi_check.values, patients[patients.surname == 'Zaitseva'].bmi.values)

[19.05582655] [19.1]


### The `treatments` Table

In [4]:
treatments

given_name       surname    auralin  ... hba1c_start  hba1c_end  hba1c_change
0     veronika      jindrová  41u - 48u  ...        7.63       7.20           NaN
1       elliot    richardson          -  ...        7.56       7.09          0.97
2     yukitaka      takenaka          -  ...        7.68       7.25           NaN
3         skye    gormanston  33u - 36u  ...        7.97       7.62          0.35
4       alissa        montez          -  ...        7.78       7.46          0.32
..         ...           ...        ...  ...         ...        ...           ...
275     albina      zetticci  45u - 51u  ...        7.93       7.73          0.20
276       john   teichelmann          -  ...        7.90       7.58           NaN
277     mathea       lillebø  23u - 36u  ...        9.04       8.67          0.37
278     vallie        prince  31u - 38u  ...        7.64       7.28          0.36
279     samúel  guðbrandsson  53u - 56u  ...        8.00       7.64          0.36

[280 rows x 7 columns]

350 patients participated in this clinical trial. None of the patients were using Novodra (a popular injectable insulin) or Auralin (the oral insulin being researched) as their primary source of insulin before.  All were experiencing elevated HbA1c levels.

All 350 patients were treated with Novodra to establish a baseline HbA1c level and insulin dose. After four weeks, which isn’t enough time to capture all the change in HbA1c that can be attributed by the switch to Auralin or Novodra:
- 175 patients switched to Auralin for 24 weeks
- 175 patients continued using Novodra for 24 weeks

`treatments` columns:
- **given_name**: the given name of each patient in the Master Patient Index that took part in the clinical trial
- **surname**: the surname of each patient in the Master Patient Index that took part in the clinical trial
- **auralin**: the baseline median daily dose of insulin from the week prior to switching to Auralin (the number before the dash) *and* the ending median daily dose of insulin at the end of the 24 weeks of treatment measured over the 24th week of treatment (the number after the dash). Both are measured in units (shortform 'u'), which is the [international unit](https://en.wikipedia.org/wiki/International_unit) of measurement and the standard measurement for insulin.
- **novodra**: same as above, except for patients that continued treatment with Novodra
- **hba1c_start**: the patient's HbA1c level at the beginning of the first week of treatment. HbA1c stands for Hemoglobin A1c. The [HbA1c test](https://depts.washington.edu/uwcoe/healthtopics/diabetes.html) measures what the average blood sugar has been over the past three months. It is thus a powerful way to get an overall sense of how well diabetes has been controlled. Everyone with diabetes should have this test 2 to 4 times per year. Measured in %.
- **hba1c_end**: the patient's HbA1c level at the end of the last week of treatment
- **hba1c_change**: the change in the patient's HbA1c level from the start of treatment to the end, i.e., `hba1c_start` - `hba1c_end`. For Auralin to be deemed effective, it must be "noninferior" to Novodra, the current standard for insulin. This "noninferiority" is statistically defined as the upper bound of the 95% confidence interval being less than 0.4% for the difference between the mean HbA1c changes for Novodra and Auralin (i.e. Novodra minus Auralin).

In [21]:
treatments.sample(5)

given_name    surname    auralin  ... hba1c_start  hba1c_end  hba1c_change
104       renzo   lucchese  53u - 64u  ...        7.54       7.17          0.37
264  chukwumoge  ogochukwu          -  ...        7.95       7.56          0.39
93       melisa       mata          -  ...        8.56       8.27          0.29
5       jasmine      sykes          -  ...        7.56       7.18          0.38
86      ananías   enríquez          -  ...        7.58       7.07          0.51

[5 rows x 7 columns]

In [22]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [31]:
treatments.auralin.isnull().sum()

0

In [33]:
treatments.novodra.isnull().sum()

0

### The `adverse_reactions` Table

In [5]:
adverse_reactions

given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache
20     anenechi       chidi               hypoglycemia
21     miłosław  wiśniewski  injection site discomfort
22        lixue       hsueh  injection site discomfort
23        merci      leroux               hypoglycemia
24         kang         mai  injection site discomfort
25       elliot  richardson               hypoglycemia
26      clinton      miller          throat irritation
27       idalia       moore               hypoglycemia
28       xiuxiu       chang               hypoglycemia
29         alex    crawford               hypoglycemia
30       monika      lončar               hypoglycemia
31       steven         roy                   headache
32      cecilie      nilsen               hypoglycemia
33    krisztina      magyar               hypoglycemia

`adverse_reactions` columns:
- **given_name**: the given name of each patient in the Master Patient Index that took part in the clinical trial and had an adverse reaction (includes both patients treated Auralin and Novodra)
- **surname**: the surname of each patient in the Master Patient Index that took part in the clinical trial and had an adverse reaction (includes both patients treated Auralin and Novodra)
- **adverse_reaction**: the adverse reaction reported by the patient

Additional useful information:
- [Insulin resistance varies person to person](http://www.tudiabetes.org/forum/t/how-much-insulin-is-too-much-on-a-daily-basis/9804/5), which is why both starting median daily dose and ending median daily dose are required, i.e., to calculate change in dose.
- It is important to test drugs and medical products in the people they are meant to help. People of different age, race, sex, and ethnic group must be included in clinical trials. This [diversity](https://www.clinicalleader.com/doc/an-fda-perspective-on-patient-diversity-in-clinical-trials-0001) is reflected in the `patients` table.
- Ensuring column names are descriptive enough is an important step in acquainting yourself with the data. 'Descriptive enough' is subjective. Ideally you want short column names (so they are easier to type and read in code form) but also fully descriptive. Length vs. descriptiveness is a tradeoff and common debate (a [similar debate](https://softwareengineering.stackexchange.com/questions/176582/is-there-an-excuse-for-short-variable-names) exists for variable names). The *auralin* and *novodra* column names are probably not descriptive enough, but you'll address that later so don't worry about that for now.

In [38]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   given_name        34 non-null     object
 1   surname           34 non-null     object
 2   adverse_reaction  34 non-null     object
dtypes: object(3)
memory usage: 944.0+ bytes


In [23]:
adverse_reactions.adverse_reaction.value_counts()

hypoglycemia                 19
injection site discomfort     6
headache                      3
nausea                        2
throat irritation             2
cough                         2
Name: adverse_reaction, dtype: int64

### Quality Issues
#### `patients`
- ZIP code is float not a string
- ZIP code has four digits sometimes
- Tim Neudorf height is 27 in instead of 72 in
- Full state names sometimes, abbreviations other times
- The *given_name* for the patient with the *patient_id* 9 is Dsvid instead of David
- Missing demographic information (address - contact columns)
- Erroneous datatypes (assigned_sex, state, zip_code, and birthdate columns)
- Multiple phone number formats
- Default John Doe data
- Multiple records for Jakobsen, Gersten, Taylor
- 48.8 kgs instead of lbs for Zaitseva weight

#### `treatments`
- Missing HbA1c changes
- The letter 'u' next to the start dose and end dose in the *auralin* and *novodra* columns
- Lowercase given names and surnames
- Missing records (280 instead of 350)
- Erroneous datatypes (auralin and novodra columns)
- Inaccurate HbA1c changes (4s mistaken as 9s)
- Nulls represented as dashes in auralin and novodra columns

#### `adverse_reactions`
- Lowercase given names and surnames




### Tidiness Issues
#### `patients`
- The contact column should be split into phone number and e-mail.

#### `treatments`
- Three variables in two columns (treatment, start dose, and end dose)

#### `adverse_reactions`

In [40]:
all_columns = pd.Series(list(patients) + list(treatments) + list(adverse_reactions))
all_columns[all_columns.duplicated()]

14    given_name
15       surname
21    given_name
22       surname
dtype: object